src/ripper/ripper.cpp

```diff
@@ -0,0 +1,305 @@
+const char kWavDataString[] = "data";
+}  // namespace
+
+Ripper::Ripper(QObject* parent)
+    : QObject(parent),
+      transcoder_(new Transcoder(this)),
+      cancel_requested_(false),
+      finished_success_(0),
+      finished_failed_(0),
+      files_tagged_(0) {
+  cdio_ = cdio_open(NULL, DRIVER_UNKNOWN);
+
+  connect(this, SIGNAL(RippingComplete()), transcoder_, SLOT(Start()));
+  connect(transcoder_, SIGNAL(JobComplete(QString, QString, bool)),
+          SLOT(TranscodingJobComplete(QString, QString, bool)));
+  connect(transcoder_, SIGNAL(AllJobsComplete()),
+          SLOT(AllTranscodingJobsComplete()));
+  connect(transcoder_, SIGNAL(LogLine(QString)), SLOT(LogLine(QString)));
+}
+
+Ripper::~Ripper() { cdio_destroy(cdio_); }
+
+void Ripper::AddTrack(int track_number, const QString& title,
+                      const QString& transcoded_filename,
+                      const TranscoderPreset& preset) {
+  if (track_number < 1 || track_number > TracksOnDisc()) {
+    qLog(Warning) << "Invalid track number:" << track_number << "Ignoring";
+    return;
+  }
+  TrackInformation track(track_number, title, transcoded_filename, preset);
+  tracks_.append(track);
+}
+
+void Ripper::SetAlbumInformation(const QString& album, const QString& artist,
+                                 const QString& genre, int year, int disc,
+                                 Song::FileType type) {
+  album_.album = album;
+  album_.artist = artist;
+  album_.genre = genre;
+  album_.year = year;
+  album_.disc = disc;
+  album_.type = type;
+}
+
+int Ripper::TracksOnDisc() const {
+  int number_of_tracks = cdio_get_num_tracks(cdio_);
+  // Return zero tracks if there is an error, e.g. no medium found.
+  if (number_of_tracks == CDIO_INVALID_TRACK) number_of_tracks = 0;
+  return number_of_tracks;
+}
+
+int Ripper::AddedTracks() const { return tracks_.length(); }
+
+void Ripper::ClearTracks() { tracks_.clear(); }
+
+bool Ripper::CheckCDIOIsValid() {
+  if (cdio_) {
+    cdio_destroy(cdio_);
+  }
+  cdio_ = cdio_open(NULL, DRIVER_UNKNOWN);
+  // Refresh the status of the cd media. This will prevent unnecessary
+  // rebuilds of the track list table.
+  if (cdio_) {
+    cdio_get_media_changed(cdio_);
+  }
+  return cdio_;
+}
+
+bool Ripper::MediaChanged() const {
+  if (cdio_ && cdio_get_media_changed(cdio_))
+    return true;
+  else
+    return false;
+}
+
+void Ripper::Start() {
+  {
+    QMutexLocker l(&mutex_);
+    cancel_requested_ = false;
+  }
+  SetupProgressInterval();
+
+  qLog(Debug) << "Ripping" << AddedTracks() << "tracks.";
+  QtConcurrent::run(this, &Ripper::Rip);
+}
+
+void Ripper::Cancel() {
+  {
+    QMutexLocker l(&mutex_);
+    cancel_requested_ = true;
+  }
+  transcoder_->Cancel();
+  RemoveTemporaryDirectory();
+  emit(Cancelled());
+}
+
+void Ripper::TranscodingJobComplete(const QString& input, const QString& output,
+                                    bool success) {
+  (*(success ? &finished_success_ : &finished_failed_))++;
```

Comment:
It is and you've commented on in [before](https://github.com/clementine-player/Clementine/pull/4113/files#r8858034). I will change it.